### PR TITLE
lodash: infer return type of flattenDeep when using _.chain

### PR DESCRIPTION
**Pull request by schmod**
Provides the correct return type when `_.chain(value)` and `_(value)` are used with `flattenDeep`.  

```ts
const a: number[][][][] = [[[[1]]]];
const b = _.chain(a)
  .flattenDeep()
  .value(); // currently typed as number[][][][] – should be number[]
```

Note that this only works with up to 4 levels of nesting, due to an internal TS design constraint that prevents types from recursing more than 5 levels deep: https://github.com/microsoft/TypeScript/issues/23024#issuecomment-377711022

---

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://lodash.com/docs/4.17.15#flattenDeep
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

File: types/lodash/ts3.1/common/array.d.ts [hunk 1]
```diff
@@ -488,6 +488,7 @@ declare module "../index" {
          */
         flatten(): T extends Many<infer U> ? CollectionChain<U> : CollectionChain<T>;
     }
+    type Flat<T> = (T extends List<infer U> ? never : T);
```
Comment (aj-r) [inline types/lodash/ts3.1/common/array.d.ts]: Why is this needed for chains but not other versions?

Also don't `infer U` if you never use it. Instead just use `any`.

File: types/lodash/ts3.1/lodash-tests.ts [hunk 2]
```diff
@@ -497,12 +497,12 @@ _.chain([1, 2, 3, 4]).unshift(5, 6); // $ExpectType CollectionChain<number>
     _.flattenDeep<number>({0: 1, 1: [2, [3, [4, 5]]], length: 2}); // $ExpectType number[]
 
     _([1, 2, 3]).flattenDeep(); // $ExpectType Collection<number>
-    _([1, [2, [3, [4, 5]]]]).flattenDeep(); // $ExpectType Collection<number | (number | (number | number[])[])[]>
-    _({0: 1, 1: [2, [3, [4, 5]]], length: 2}).flattenDeep(); // $ExpectType Collection<number | (number | (number | number[])[])[]>
+    _([1, [2, [3, [4, 5]]]]).flattenDeep(); // $ExpectType Collection<number>
```
Comment (aj-r) [inline types/lodash/ts3.1/lodash-tests.ts]: It might be nice to test some other cases too. Specifically I'm worried about a case like `_([[[1, 2]]])` resulting in `Collection<never>`

**Review by aj-r — CHANGES_REQUESTED** [review]

File: types/lodash/ts3.1/lodash-tests.ts (hunk 2 above)
Comment (schmod) [reply types/lodash/ts3.1/lodash-tests.ts]: I've added this test case.

File: types/lodash/ts3.1/common/array.d.ts (hunk 1 above)
Comment (schmod) [reply types/lodash/ts3.1/common/array.d.ts]: I've switched `infer U` to `any`.  

I didn't realize that the static version wasn't working until I noticed that the tests were all asserting the final type via a generic.  I've added this workaround to the `LodashStatic` variant of `_.flattenDeep()` and updated the tests to add cases where the explicit generics are not provided.

File: types/lodash/ts3.1/lodash-tests.ts [hunk 3]
```diff
@@ -492,19 +492,30 @@ _.chain([1, 2, 3, 4]).unshift(5, 6); // $ExpectType CollectionChain<number>
 
 // _.flattenDeep
 {
+    _.flattenDeep([1, 2, 3]); // $ExpectType number[]
+    _.flattenDeep([[1, 2, 3]]); // $ExpectType number[]
+    _.flattenDeep([1, [2, [3, [4, 5]]]]); // $ExpectType number[]
+    _.flattenDeep([1, [2, [3, [4, 5]]]]); // $ExpectType number[]
```
Comment (aj-r) [inline types/lodash/ts3.1/lodash-tests.ts]: Duplicate?

**Review by aj-r — CHANGES_REQUESTED** [review]
Looks good, just 1 nitpick

**Review by aj-r — APPROVED** [review]

**Comment by stocksr** [comment]
Not sure if this is the best place for this but this release broke my code when using arrays of strings 
I am getting a return type of `never[]`  from a input type of `(string | string[])[]`

Looking at this PR I think adding and fixing these test cases might be the way forward.
```
_.flattenDeep(['x']); // $ExpectType string[]
_.flattenDeep(['x',['y']]); // $ExpectType string[]
```

**Comment by aj-r** [comment]
Weird... maybe it's confused because `string` also implements `ArrayLike<string>`? If that's the case, then we may need to add a special overload for strings.

In the meantime, does this workaround work?

`_.flattenDeep<string>(['x',['y']]);`

**Comment by aj-r** [comment]
Ah I think I see the problem. We probably need to change:

` type Flat<T> = (T extends List<any> ? never : T);`

to:

` type Flat<T> = T extends string ? T : (T extends List<any> ? never : T);`

**Comment by stocksr** [comment]
Hi,  Finally got back around to looking at this again.

> In the meantime, does this workaround work?
> 
> `_.flattenDeep<string>(['x',['y']]);`

No sorry did not work

> Ah I think I see the problem. We probably need to change:
> 
> ` type Flat<T> = (T extends List<any> ? never : T);`
> 
> to:
> 
> ` type Flat<T> = T extends string ? T : (T extends List<any> ? never : T);`


made this change in my local node_modules folder and this works great.

is there an issue tracker I should move this to in-order to get it changed?

**Comment by aj-r** [comment]
Sure, you can create a bug in DefinitelyTyped with "lodash" in the title. Or better yet, create a PR that implements this change (it will get fixed much faster that way).
